goalc/emitter/Register.h

```diff
@@ -60,6 +60,78 @@ enum X86_REG : s8 {
   XMM15,  // saved
 };
 
+// TODO - i think it'll be better to make some sort of abstraction
+// mapping between x86 and arm, btu just using this enum as a place to prototype
+// the registers to use.
```

Comment:
I think we should avoid an abstraction over registers, and the existing abstractions within the compiler are enough. I would keep this enum (maybe split it for simd/gprs), and add ArmGpr and ArmSimd to HwRegKind.

For frontend compiler stuff, we can modify `make_gpr()`/`make_fpr()` to return appropriately types for the target architecture. In places where we currently specify x86 register constraints in the front end, we'll need to make ARM specific versions. It won't be possible to abstract these places. (but there are only a few).

For instruction/code generation stuff, registers are treated as just a `Kind`, `int`, where the `Kind` tells you xmm/gpr/(arm_gpr)/(arm_simd), and the second int is the "hardware id" that actually goes in the instruction.

We'll need to eventually write GOAL code that refers to a specific register to implement stuff like switching GOAL threads.  There are also weird ways that we'll need to generate code/constrain registers differently. (for example, x86 variable shifts require that shift amount is in rcx, and division has some weird requirements too).

For GOAL code outside of the kernel, we never specify register mapping. So the existing `rlet`s in code aren't specific to x86.

Reply:
Yeah my primary concern from doing a few searches was how there's a decent amount of code that refers to explicit registers, atleast by name.  Ranging from the debugger code, the IR code, and obviously the kernels.  But I havn't dug into any of them in detail yet -- I just don't want to make the code a mess of ifdefs.  

It certainly makes sense in some spots like the kernel, but I was imagining nothing more than an abstraction like "give me the register for arg0" which either returns `rdi` or `x0` instead of that being communicated by comments.

But when I look into fixing up the related code, it might not even be relevant or worth it.

Reply:
> , but I was imagining nothing more than an abstraction like "give me the register for arg0" which either returns rdi or x0 instead of that being communicated by comments.

Yeah, that sounds great to me.  I think the key detail is that the `Register` type should still hold actual real hardware registers (`rdi` or `x0`), rather than some generic register representation. The compiler is designed so that we only `Register` when we really care a specific register. 

The specific example you mentioned is mostly already done:
https://github.com/open-goal/jak-project/blob/master/goalc/compiler/compilation/Function.cpp#L185. We'd need to add an ARM case, but we already have a function that returns `vector<Register>` for all the arguments/return value in a function call. 
(due to 128-bit arguments/return values, the actual registers used depend on the type of the function)

>  there's a decent amount of code that refers to explicit registers, atleast by name.

Maybe I'm missing something, but I don't think there's too much here. I _think_ I've done a decent job of only doing this when absolutely needed. Stuff like this:
https://github.com/open-goal/jak-project/blob/master/goalc/compiler/compilation/Math.cpp#L465
is just a requirement of x86 and it will be different on ARM (which needs no constraint here, I think you can use any register for integer division).